CDAP-14260 endpoint to get the latest run for multiple programs

albertshau opened the pull request:
Added an endpoint to get the latest run for a batch of programs.
The endpoint follows the same format as the other batch endpoints.

Did some code cleanup to move runs lookup for a program from the
handler into the ProgramLifecycleService. This refactoring fixed
a bug where auth checks were not being applied to the program
runs endpoint.

albertshau committed:
CDAP-14260 endpoint to get the latest run for multiple programs

Added an endpoint to get the latest run for a batch of programs.
The endpoint follows the same format as the other batch endpoints.

Did some code cleanup to move runs lookup for a program from the
handler into the ProgramLifecycleService. This refactoring fixed
a bug where auth checks were not being applied to the program
runs endpoint.

albertshau commented:
https://builds.cask.co/browse/CDAP-DUT6637

yaojiefeng commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java:
```diff
@@ -203,6 +204,32 @@ public long getProgramRunCount(ProgramId programId) throws Exception {
   }
 
+  /**
+   * Get the latest run within the specified start and end times for the specified program.
```
nit: `Get the latest runs`

yaojiefeng commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java:
```diff
@@ -203,6 +204,32 @@ public long getProgramRunCount(ProgramId programId) throws Exception {
+   */
+  public List<RunRecord> getRuns(ProgramId programId, ProgramRunStatus programRunStatus,
+                                 long start, long end, int limit) throws Exception {
+    AuthorizationUtil.ensureAccess(programId, authorizationEnforcer, authenticationContext.getPrincipal());
```
this will make a REST call for each program to make the auth check, there is an `isVisible()` method in `AuthorizationEnforcer` which takes a collection of the entity ids and do the auth check in batch. I think what we did for run count is to have the method `getProgramRunCounts` take a list of programs and doing all the operations in batch instead of starting the transaction for each operation.

yaojiefeng commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/ProgramLifecycleService.java:
```diff
@@ -203,6 +204,32 @@ public long getProgramRunCount(ProgramId programId) throws Exception {
+  public List<RunRecord> getRuns(ProgramId programId, ProgramRunStatus programRunStatus,
+                                 long start, long end, int limit) throws Exception {
+    AuthorizationUtil.ensureAccess(programId, authorizationEnforcer, authenticationContext.getPrincipal());
+    ProgramSpecification programSpec = getProgramSpecificationWithoutAuthz(programId);
```
I think we are having this check existence and then perform the get in many places, I think it is better we can combine these in one transaction.

albertshau committed:
CDAP-14260 batch program run history calls

To avoid generating a bunch of unnecessary transactions, batch
the calls to get program history.

yaojiefeng commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java:
```diff
@@ -915,6 +918,57 @@ public long getProgramRunCount(ProgramId programId) throws NotFoundException {
+      AppMetadataStore appMetadataStore = getAppMetadataStore(context);
+
+      for (ProgramId program : programs) {
+        ApplicationMeta appMeta = appMetadataStore.getApplication(program.getParent());
```
You can use `getApplicationsForAppIds` in `AppMetadatStore` to batch get the app meta. It will return a map of app id to app meta.

albertshau replied:
done

yaojiefeng commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java:
```diff
@@ -915,6 +918,57 @@ public long getProgramRunCount(ProgramId programId) throws NotFoundException {
+        }
+
+        ApplicationSpecification appSpec = appMeta.getSpec();
+        ProgramType type = program.getType();
```
You can use `getExistingAppProgramSpecification()` in this class to get the program spec

albertshau replied:
good suggestion, should have looked around for similar functionality.

yaojiefeng commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/store/DefaultStore.java:
```diff
@@ -915,6 +918,57 @@ public long getProgramRunCount(ProgramId programId) throws NotFoundException {
+          result.add(new ProgramHistory(program, Collections.emptyList(), new ProgramNotFoundException(program)));
+          continue;
+        }
+
```
nit: extra new line

albertshau replied:
done

albertshau committed:
re-use existing methods

yaojiefeng reviewed (approved):
lgtm
